polymetis/polymetis/python/polysim/envs/experimental/bullet_manipulator.py

```diff
@@ -0,0 +1,164 @@
+                targetValue=self.cfg.rest_pose[i],
+                targetVelocity=0,
+            )
+
+        # Initialize states
+        self.arm_state = polymetis_pb2.RobotState()
+        self.arm_state.prev_joint_torques_computed[:] = np.zeros(7)
+        self.arm_state.prev_joint_torques_computed_safened[:] = np.zeros(7)
+        self.arm_state.motor_torques_measured[:] = np.zeros(7)
+        self.arm_state.motor_torques_external[:] = np.zeros(7)
+
+        self.arm_state.prev_command_successful = True
+        self.arm_state.error_code = 0
+
+        self.gripper_state = polymetis_pb2.GripperState()
+
+    def get_arm_state(self) -> polymetis_pb2.RobotState:
+        # Timestamp
+        self.arm_state.timestamp.GetCurrentTime()
+
+        # Joint pos & vel
+        joint_cur_states = self.sim.getJointStates(
+            self.robot_id, self.cfg.controlled_joints
+        )
+        self.arm_state.joint_positions[:] = [joint_cur_states[i][0] for i in range(7)]
+        self.arm_state.joint_velocities[:] = [joint_cur_states[i][1] for i in range(7)]
+
+        return self.arm_state
+
+    def get_gripper_state(self) -> polymetis_pb2.GripperState:
+        joint_cur_states = self.sim.getJointStates(
+            self.robot_id, self.cfg.gripper.controlled_joints
+        )
+        self.gripper_state.width = float(
+            joint_cur_states[0][0] + joint_cur_states[1][0]
+        )
+        self.gripper_state.is_grasped = False  # TODO
+        self.gripper_state.is_moving = np.all(
+            [
+                abs(joint_cur_states[i][1]) < self.cfg.gripper.moving_threshold
+                for i in range(2)
+            ]
+        )
+
+        return self.gripper_state
+
+    def apply_arm_control(self, cmd: polymetis_pb2.TorqueCommand):
+        # Extract torques
+        commanded_torques = np.array(list(cmd.joint_torques))
+
+        # Compute grav comp
+        joint_pos = list(self.arm_state.joint_positions)
+        finger_pos = [self.gripper_state.width / 2.0] * 2
+        grav_comp_torques = self.sim.calculateInverseDynamics(
+            self.robot_id,
+            joint_pos + finger_pos,
+            [0] * 9,
+            [0] * 9,
+        )[:7]
+
+        # Set sim torques
+        applied_torques = commanded_torques + grav_comp_torques
+        self.sim.setJointMotorControlArray(
+            bodyIndex=self.robot_id,
+            jointIndices=self.cfg.controlled_joints,
+            controlMode=pybullet.TORQUE_CONTROL,
+            forces=applied_torques,
+        )
+
+        # Populate torques in state
+        self.arm_state.prev_joint_torques_computed[:] = commanded_torques
+        self.arm_state.prev_joint_torques_computed_safened[:] = commanded_torques
+        self.arm_state.motor_torques_measured[:] = applied_torques
+        self.arm_state.motor_torques_external[:] = np.zeros_like(applied_torques)
+
+        self.arm_state.prev_command_successful = True
+        self.arm_state.error_code = 0
```

Comment:
When is error not 0?

Reply:
The field exists to identify which of the numerous things that can go wrong on hardware. In the case of sim I'm just setting it to 0. Can probably omit this line since the protobuf message would treat an unassigned int field as 0.